include/crab/domains/mru_region_domain.hpp

```diff
@@ -0,0 +1,2270 @@
+  // the actual variable's type must be inferred dynamically (as part
+  // of the abstract domain). In this case, it will return true if the
+  // type inference done by the abstract domain succeeds (type is
+  // inferred different from unknown region).
+  bool has_dynamic_type(const variable_t &v) const {
+    if (!v.get_type().is_unknown_region()) {
+      return true;
+    }
+
+    if (crab_domain_params_man::get().region_skip_unknown_regions()) {
+      return false;
+    } else {
+      type_value dyn_ty = m_rgn_type_dom[v];
+      if (dyn_ty.is_top() || dyn_ty.is_bottom()) {
+        return false;
+      }
+      return (!dyn_ty.get().is_unknown_region());
+    }
+  }
+
+  // if has_dynamic_type(v) returns true then it returns the type,
+  // otherwise it will fail.
+  variable_type get_dynamic_type_or_fail(const variable_t &v) const {
+    assert(has_dynamic_type(v));
+
+    if (!v.get_type().is_unknown_region()) {
+      return v.get_type();
+    }
+
+    type_value dyn_ty = m_rgn_type_dom[v];
+    if (dyn_ty.is_top() || dyn_ty.is_bottom()) {
+      CRAB_ERROR("get_dynamic_type_or_fail cannot be called on top or bottom");
+    }
+    if (dyn_ty.get().is_unknown_region()) {
+      CRAB_ERROR("get_dynamic_type_or_fail should not return unknown region");
+    }
+    return dyn_ty.get();
+  }
+
+  static void ERROR_IF_NOT_REGION(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_region()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not a region at line ", line);
+    }
+  }
+
+  static void ERROR_IF_ARRAY_REGION(const variable_t &v, unsigned line) {
+    if (v.get_type().is_array_region()) {
+      CRAB_ERROR(v, ":", v.get_type(), " cannot contain an array at line ",
+                 line);
+    }
+  }
+
+  static void ERROR_IF_NOT_REF(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_reference()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not a reference at line ", line);
+    }
+  }
+
+  static void ERROR_IF_NOT_INT(const variable_t &v, unsigned line) {
+    if (!v.get_type().is_integer()) {
+      CRAB_ERROR(v, ":", v.get_type(), " is not an integer at line ", line);
+    }
+  }
+
+  /**------------------ End helper method definitions ------------------**/
+
+  /**------------- Begin MRU helper method definitions -------------------**/
+  // perform a similiar work for cache.lines.constraint(cache.regs)
+  // This one requires to be renaming because mem base dom used different base
+  // vars
+  base_abstract_domain_t cache_lines_regs_reduce() {
+    base_abstract_domain_t res(m_cache_lines);
+    auto csts = m_regs.to_linear_constraint_system();
+    res += csts;
+    return res;
+  }
+
+  // get all used Crab IR vars by region dom
+  void vars(variable_vector_t &used_vars) const {
+    used_vars.reserve(m_var_map.size());
+    for (auto &kv : m_var_map) {
+      used_vars.push_back(kv.first);
+    }
+  }
+
+  void region_vars_in_mem(variable_vector_t &used_vars) const {
```

Comment:
This is inefficient. You can use `boost::transform_iterator` so that you can iterate e.g., over a map but the caller only iterates over the first or the second elements. This avoids makes memory allocations and copies. Of course, you need to adapt the callers of `region_vars_in_mem`. For instance, `vec_minus`  uses it. So you can adapt `vec_minus` to take a pair of iterators rather than a vector. LMK if this is not clear.